nft-minter/src/nft_module.rs

```diff
@@ -0,0 +1,211 @@
+            "Invalid price token"
+        );
+
+        let is_new_brand = self.registered_brands().insert(brand_id.clone());
+        require!(is_new_brand, "Brand already exists");
+
+        let id_offset = self.last_item_id().update(|last_id| {
+            let prev_last_id = *last_id;
+            *last_id += max_nfts;
+            require!(*last_id > prev_last_id, "ID overflow!");
+
+            prev_last_id
+        });
+        let brand_info = BrandInfo {
+            token_display_name: token_display_name.clone(),
+            media_type,
+            id_offset,
+            royalties,
+            mint_start_epoch,
+            mint_price_token_id,
+            mint_price_amount,
+        };
+        self.brand_info(&brand_id).set(&brand_info);
+        self.available_ids(&brand_id).set_initial_len(max_nfts);
+
+        if !tags.is_empty() {
+            self.tags_for_brand(&brand_id).set(&tags.to_vec());
+        }
+
+        self.nft_token(&brand_id).issue(
+            EsdtTokenType::NonFungible,
+            payment_amount,
+            token_display_name,
+            token_ticker,
+            0,
+            Some(self.callbacks().issue_callback(brand_id)),
+        );
+    }
+
+    #[callback]
+    fn issue_callback(
+        &self,
+        brand_id: BrandId<Self::Api>,
+        #[call_result] result: ManagedAsyncCallResult<TokenIdentifier>,
+    ) {
+        match result {
+            ManagedAsyncCallResult::Ok(token_id) => {
+                self.nft_token(&brand_id).set_token_id(&token_id);
+            }
+            ManagedAsyncCallResult::Err(_) => {
+                self.brand_info(&brand_id).clear();
+                self.tags_for_brand(&brand_id).clear();
+                self.available_ids(&brand_id).clear_len();
+                let _ = self.registered_brands().swap_remove(&brand_id);
+            }
+        }
+    }
+
+    #[endpoint(setLocalRoles)]
+    fn set_local_roles(&self, brand_id: BrandId<Self::Api>) {
+        self.nft_token(&brand_id)
+            .set_local_roles(&[EsdtLocalRole::NftCreate], None);
+    }
+
+    #[payable("*")]
+    #[endpoint(buyRandomNft)]
+    fn buy_random_nft(&self, brand_id: BrandId<Self::Api>) {
+        require!(
+            self.registered_brands().contains(&brand_id),
+            INVALID_BRAND_ID_ERR_MSG
+        );
+
+        let brand_info: BrandInfo<Self::Api> = self.brand_info(&brand_id).get();
+        let payment: EsdtTokenPayment<Self::Api> = self.call_value().payment();
+        require!(
+            payment.token_identifier == brand_info.mint_price_token_id
+                && payment.amount == brand_info.mint_price_amount,
+            "Invalid payment"
+        );
+
+        self.add_mint_payment(payment.token_identifier, payment.amount);
+
+        let caller = self.blockchain().get_caller();
+        self.send_random_nft(&caller, &brand_id, &brand_info);
+    }
+
+    #[endpoint(givawayNft)]
```

Comment:
Bulk giveaway would be nice. Args would look like MultiValueEncoded<MultiValue2<ManagedAddress, usize>>

Reply:
Done.